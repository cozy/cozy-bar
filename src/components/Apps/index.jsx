import React, { Component } from 'react'
import PropTypes from 'prop-types'

import AppsContent from 'components/Apps/AppsContent'
import AppNavButtons from 'components/Apps/AppNavButtons'

class Apps extends Component {
  constructor(props) {
    super(props)
    this.state = {
      opened: false
    }
  }

  componentDidMount() {
    document.body.addEventListener('click', this.onClickOutside)
    this.modalContainer = document.getElementById('cozy-bar-modal-dom-place')
  }

  componentWillUnmount() {
    document.body.removeEventListener('click', this.onClickOutside)
  }

  onClickOutside = event => {
    if (this.state.opened) {
      // if it's not a cozy-bar nav popup, close the opened popup
      if (
        !this.rootRef.contains(event.target) &&
        !this.modalContainer.contains(event.target)
      ) {
        this.setState({ opened: false })
        event.stopPropagation()
      }
    }
  }

  toggleMenu = () => {
    this.setState({ opened: !this.state.opened })
  }

  // data-tutorial attribute allows to be targeted in an application tutorial
  render() {
    const {
      appName,
      appNamePrefix,
      appSlug,
      iconPath,
      isPublic,
      isInvertedTheme
    } = this.props
    const { opened } = this.state
    return (
      <nav
        className="coz-nav coz-nav-apps"
        ref={ref => {
          this.rootRef = ref
        }}
      >
        <AppNavButtons
          appName={appName}
          appNamePrefix={appNamePrefix}
          appSlug={appSlug}
          iconPath={iconPath}
          handleClick={this.toggleMenu}
          opened={opened}
          isPublic={isPublic}
          isInvertedTheme={isInvertedTheme}
        />
        <div
          className="coz-nav-pop coz-nav-pop--apps"
          id="coz-nav-pop--apps"
          aria-hidden={!opened}
        >
          <AppsContent isInvertedTheme={isInvertedTheme} />
        </div>
      </nav>
    )
  }
}

Apps.propTypes = {
  appName: PropTypes.string,
  appNamePrefix: PropTypes.string,
  appSlug: PropTypes.string,
  iconPath: PropTypes.string,
  isPublic: PropTypes.bool,
  isInvertedTheme: PropTypes.bool
}

export default Apps
