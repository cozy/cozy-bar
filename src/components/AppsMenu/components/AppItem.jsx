import React from 'react'
import get from 'lodash/get'

import { models } from 'cozy-client'
import Buttons from 'cozy-ui/transpiled/react/Buttons'
import AppIcon from 'cozy-ui/transpiled/react/AppIcon'
import AppLinker from 'cozy-ui/transpiled/react/AppLinker'
import Typography from 'cozy-ui/transpiled/react/Typography'
import cx from 'classnames'

import { appShape } from 'proptypes/index'
import stack from 'lib/stack'

import styles from 'styles/apps-menu.styl'

const getAppDisplayName = get(models, 'applications.getAppDisplayName', app => {
  return app.namePrefix && app.namePrefix.toLowerCase() !== 'cozy'
    ? `${app.namePrefix} ${app.name}`
    : app.name
})

export const AppItem = ({ app }) => {
  const appName = getAppDisplayName(app)

  return (
    <AppLinker href={app.href || ''} app={app}>
      {({ onClick, href }) => {
        return (
          <Buttons
            height="auto"
            component="a"
            variant="text"
            href={href}
            title={appName}
            onClick={onClick}
            className={cx(styles['apps-menu-grid-item-wrapper'], 'u-bdrs-5')}
            label={
              <div className={styles['apps-menu-grid-item']}>
                <AppIcon app={app} key={app.slug} {...stack.get.iconProps()} />
                <Typography variant="caption" noWrap className="u-mt-half">
                  {appName}
                </Typography>
              </div>
            }
          />
        )
      }}
    </AppLinker>
  )
}

AppItem.propTypes = {
  app: appShape.isRequired
}

export default AppItem
