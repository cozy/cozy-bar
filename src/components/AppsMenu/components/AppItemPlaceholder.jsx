import React from 'react'

import Skeleton from 'cozy-ui/transpiled/react/Skeleton'

export const AppItemPlaceholder = () => {
  return (
    <Skeleton
      variant="rect"
      width={48}
      height={48}
      className="u-m-1 u-bdrs-5"
    />
  )
}

export default AppItemPlaceholder
