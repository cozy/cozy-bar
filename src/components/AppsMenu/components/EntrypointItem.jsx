import React from 'react'

import { useClient, generateWebLink } from 'cozy-client'

import { makeStyles } from 'cozy-ui/transpiled/react/styles'
import Buttons from 'cozy-ui/transpiled/react/Buttons'
import Typography from 'cozy-ui/transpiled/react/Typography'
import cx from 'classnames'

import styles from 'styles/apps-menu.styl'
import useI18n from 'components/useI18n'

const useStyles = makeStyles(() => {
  return {
    text: {
      lineHeight: '22.5px',
      fontSize: '12px',
      fontWeight: 400
    }
  }
})

export const EntrypointItem = ({ entrypoint }) => {
  const client = useClient()
  const classes = useStyles()
  const { lang } = useI18n()

  const cozyUrl = client.getStackClient().uri
  const { subdomain: subDomainType } = client.getInstanceOptions()

  const entrypointUrl = generateWebLink({
    cozyUrl,
    subDomainType,
    slug: entrypoint.slug,
    pathname: '/',
    hash: entrypoint.hash
  })

  const title = entrypoint.title[lang] || entrypoint.title['en']

  return (
    <Buttons
      height="auto"
      component="a"
      target="_blank"
      variant="text"
      href={entrypointUrl}
      title={title}
      className={cx(styles['apps-menu-grid-item-wrapper'], 'u-bdrs-5')}
      label={
        <div className={styles['apps-menu-grid-item']}>
          <div className={styles['apps-menu-grid-item-icon']}>
            <img
              className="u-bdrs-5"
              src={`data:image/svg+xml;base64,${entrypoint.icon}`}
              alt=""
              width={42}
              height={42}
            />
          </div>
          <Typography noWrap className={classes.text}>
            {title}
          </Typography>
        </div>
      }
    />
  )
}

export default EntrypointItem
