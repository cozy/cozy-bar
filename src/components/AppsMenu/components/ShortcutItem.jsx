import React from 'react'

import { useClient, useFetchShortcut } from 'cozy-client'
import { splitFilename, getShortcutImgSrc } from 'cozy-client/dist/models/file'

import { makeStyles } from 'cozy-ui/transpiled/react/styles'
import Buttons from 'cozy-ui/transpiled/react/Buttons'
import Typography from 'cozy-ui/transpiled/react/Typography'
import cx from 'classnames'

import styles from 'styles/apps-menu.styl'

const useStyles = makeStyles(() => {
  return {
    text: {
      lineHeight: '22.5px',
      fontSize: '12px',
      fontWeight: 400
    }
  }
})

export const ShortcutItem = ({ shortcut }) => {
  const client = useClient()
  const classes = useStyles()
  const { shortcutInfos, shortcutImg } = useFetchShortcut(client, shortcut._id)

  const { filename } = splitFilename(shortcut)
  const url = shortcutInfos?.data?.url || '#'

  const shortcutImgSrc = getShortcutImgSrc(shortcut)

  return (
    <Buttons
      height="auto"
      component="a"
      target="_blank"
      variant="text"
      href={url}
      title={filename}
      className={cx(styles['apps-menu-grid-item-wrapper'], 'u-bdrs-5')}
      label={
        <div className={styles['apps-menu-grid-item']}>
          <div className={styles['apps-menu-grid-item-icon']}>
            {shortcutImgSrc ? (
              <img className="u-bdrs-5 u-w-100" src={shortcutImgSrc} alt="" />
            ) : (
              <img className="u-bdrs-5 u-w-100" src={shortcutImg} alt="" />
            )}
          </div>
          <Typography noWrap className={classes.text}>
            {filename}
          </Typography>
        </div>
      }
    />
  )
}

export default ShortcutItem
