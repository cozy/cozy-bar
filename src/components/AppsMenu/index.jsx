import React, { useRef, useState } from 'react'

import AppsMenuContent from 'components/AppsMenu/AppsMenuContent'

import IconButton from 'cozy-ui/transpiled/react/IconButton'
import Icon from 'cozy-ui/transpiled/react/Icon'
import Menu from 'cozy-ui/transpiled/react/Menu'
import MosaicIcon from 'cozy-ui/transpiled/react/Icons/Mosaic'
import { useBreakpoints } from 'cozy-ui/transpiled/react/providers/Breakpoints'
import { ConfirmDialog } from 'cozy-ui/transpiled/react/CozyDialogs'
import { makeStyles } from 'cozy-ui/transpiled/react/styles'

const useStyles = makeStyles({
  root: {
    '& .dialogContentInner': {
      '& .dialogContentWrapper': {
        paddingBottom: '0 !important'
      }
    }
  }
})

const AppsMenu = () => {
  const [isOpen, setOpen] = useState(false)
  const containerRef = useRef()
  const buttonRef = useRef()

  const { isMobile } = useBreakpoints()

  const toggleMenu = () => {
    setOpen(!isOpen)
  }

  const styles = useStyles()

  return (
    <nav ref={containerRef}>
      <IconButton ref={buttonRef} onClick={toggleMenu} className="u-p-half">
        <Icon icon={MosaicIcon} size="18" />
      </IconButton>
      {isMobile ? (
        <ConfirmDialog
          open={isOpen}
          onClose={toggleMenu}
          content={<AppsMenuContent />}
          componentsProps={{
            dialogContent: {
              classes: styles
            }
          }}
          classes={{
            paper: 'u-bdrs-7'
          }}
        />
      ) : (
        <Menu
          open={isOpen}
          anchorEl={buttonRef.current}
          container={containerRef.current}
          getContentAnchorEl={null}
          onClose={toggleMenu}
          anchorOrigin={{
            vertical: 'bottom',
            horizontal: 'left'
          }}
          transformOrigin={{
            vertical: -10,
            horizontal: 0
          }}
          classes={{
            paper: 'u-bdrs-7'
          }}
        >
          <AppsMenuContent />
        </Menu>
      )}
    </nav>
  )
}

export default AppsMenu
