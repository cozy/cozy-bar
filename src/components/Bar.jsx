/* global __TARGET__ */
/* global __PIWIK_TRACKER_URL__  __PIWIK_SITEID__ __PIWIK_DIMENSION_ID_APP__ */

import React, { Component } from 'react'
import { translate } from '../lib/I18n'
import { shouldEnableTracking, getTracker, configureTracker } from 'cozy-ui/react/helpers/tracker'

import Drawer from './Drawer'
import Nav from './Nav'
import Claudy from './Claudy'

class Bar extends Component {
  constructor (props, context) {
    super(props)
    this.store = context.store
    this.state = {
      enableClaudy: null, // no claudy by default
      fireClaudy: false, // true to fire claudy (used by the drawer)
      claudyOpened: false,
      drawerVisible: false,
      usageTracker: null
    }
    this.toggleDrawer = this.toggleDrawer.bind(this)
  }

  async componentWillMount () {
    const enableClaudy = await this.store.shouldEnableClaudy()
    this.setState({ enableClaudy })
  }

  componentDidMount () {
    // if tracking enabled, init the piwik tracker
    if (shouldEnableTracking()) {
      const trackerInstance = getTracker(__PIWIK_TRACKER_URL__, __PIWIK_SITEID__, false, false)
      configureTracker({
        appDimensionId: __PIWIK_DIMENSION_ID_APP__,
        app: 'Cozy Bar',
        heartbeat: 0
      })
      this.setState({ usageTracker: trackerInstance })
    }
  }

  toggleDrawer () {
    // don't allow to toggle the drawer if claudy opened or is opening
    if (this.state.claudyOpened || this.state.fireClaudy) return
    const drawerVisible = !this.state.drawerVisible
    // don't wait for transitionend if displaying
    if (drawerVisible) this.props.onDrawer(drawerVisible)
    this.setState({ drawerVisible })
  }

  toggleClaudy (isFromDrawer = false) {
    if (!this.state.enableClaudy) return
    const { usageTracker, claudyOpened } = this.state
    if (isFromDrawer && !claudyOpened) { // if opened from drawer
      // reset to toggle via the Claudy component
      return this.setState({fireClaudy: true})
    }
    if (this.state.fireClaudy) this.setState({fireClaudy: false})
    if (usageTracker) {
      usageTracker.push([
        'trackEvent',
        'Claudy',
        claudyOpened ? 'close' : 'open',
        'claudy'
      ])
    }
    this.setState({ claudyOpened: !claudyOpened })
  }

  render () {
    const { t, lang, appName,
      appEditor, iconPath, replaceTitleOnMobile,
      onDrawer, isPublic } = this.props
    const { usageTracker, claudyOpened,
      enableClaudy, drawerVisible, fireClaudy } = this.state
    return (
      <div class='coz-bar-container'>
        <h1 lang={lang} class={`coz-bar-title ${replaceTitleOnMobile ? 'coz-bar-hide-sm' : ''}`}>
          <img class='coz-bar-hide-sm' src={iconPath} width='32' />
          {appEditor && <span class='coz-bar-hide-sm'>{appEditor} </span>}
          <strong>{appName}</strong>
          <sup class='coz-bar-hide-sm coz-bar-beta-status'>{t('beta')}</sup>
        </h1>
        {__TARGET__ !== 'mobile' && !isPublic &&
          <div class='coz-bar-flex-container'>
            <button class='coz-bar-burger' onClick={this.toggleDrawer} data-icon='icon-hamburger'>
              <span class='coz-bar-hidden'>{t('drawer')}</span>
            </button>
            <Drawer visible={drawerVisible} onClose={this.toggleDrawer} onClaudy={(enableClaudy && (() => this.toggleClaudy(true))) || false} isClaudyLoading={fireClaudy} drawerListener={() => onDrawer(this.state.drawerVisible)} />
            <Nav />
            {enableClaudy &&
              <Claudy
                usageTracker={usageTracker}
                fireClaudy={fireClaudy}
                onToggle={() => this.toggleClaudy(false)}
                opened={claudyOpened}
              />
            }
          </div>
        }
      </div>
    )
  }
}

export default translate()(Bar)
