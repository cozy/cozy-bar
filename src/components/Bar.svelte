<button class='coz-bar-burger' on:click='toggleDrawer()' data-icon='icon-hamburger'>
  {{t('menu')}}
</button>

<h1 class='coz-bar-title'>
  <img class='coz-bar-hide-sm' src='{{iconPath}}' width='32' />
  <span class='coz-bar-hide-sm'>cozy </span><strong>{{appName}}</strong>
</h1>

<hr class='coz-sep-flex' />

<Navigation sections='{{sections}}' on:open='onPopOpen(event.panel)' />

<Drawer ref:drawer content='{{config.apps.length?config.apps[0]:false}}' footer='{{sections[1].items}}' />

<script>
  import { t } from '../lib/i18n'
  import stack from '../lib/stack'

  import Navigation from './Navigation'
  import Drawer from './Drawer'

  import MENU_CONFIG from '../config/menu'

  async function updateAppsItems () {
    try {
      let apps = await stack.get.apps()
      apps = [await apps.map(app => {
        return {
          label: app.attributes.name,
          external: true,
          l10n: false,
          href: `https://${app.attributes.slug}.cozy.local/`,
          icon: `https://${app.attributes.slug}.cozy.local/${app.attributes.icon}`
        }
      })]
      this.set({ config: Object.assign({}, this.get('config'), {apps}) })
    } catch (e) {
      console.error(e.message)
    }
  }

  async function updateDiskUsage () {
      let currentDiskUsage
    try {
      let data = await stack.get.diskUsage()
      currentDiskUsage = parseInt(data.attributes.used)
    } catch (e) {
      console.error(e.message)
      currentDiskUsage = { error: e.name }
    }
    // copy settings section to update storage item
    const newSettings = this.get('config').settings.slice()
    newSettings.forEach(section => {
      let storageItem = section.find(item => item.label === 'storage')
      if (storageItem) {
        storageItem.currentDiskUsage = currentDiskUsage
      }
    })
    this.set({ config: Object.assign({}, this.get('config'), {settings: newSettings}) })
  }

  export default {
    data() {
      return {
        config: MENU_CONFIG
      }
    },

    computed: {
      sections: config => {
        return [{
          label: 'apps',
          icon: 'icon-cube',
          items: config.apps
        }, {
          label: 'settings',
          icon: 'icon-cog',
          items: config.settings
        }]
      }
    },

    components: {
      Navigation,
      Drawer
    },

    helpers: { t },

    methods: {
      toggleDrawer() {
        this.refs.drawer.set({folded: false})
        updateAppsItems.call(this)
      },
      onPopOpen (panel) {
        if (panel === 'apps') {
          updateAppsItems.call(this)
        } else if (panel === 'settings') {
          updateDiskUsage.call(this)
        }
      }
    }
  }
</script>
