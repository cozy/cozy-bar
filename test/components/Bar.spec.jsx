import React from 'react'
import { Bar } from 'components/Bar'
import { shallow } from 'enzyme'
import { isMobileApp } from 'cozy-device-helper'
import toJson from 'enzyme-to-json'
jest.mock('cozy-device-helper', () => ({
  ...require.requireActual('cozy-device-helper'),
  isMobileApp: jest.fn()
}))

describe('Bar', () => {
  let props
  beforeEach(() => {
    isMobileApp.mockReturnValue(false)
    props = {
      fetchContext: jest.fn().mockResolvedValue({}),
      fetchApps: jest.fn().mockResolvedValue([]),
      fetchSettingsData: jest.fn().mockResolvedValue({}),
      t: x => x
    }
  })

  it('should fetch data when mounted', () => {
    const root = shallow(<Bar {...props} />)
    expect(props.fetchContext).toHaveBeenCalled()
    expect(props.fetchApps).toHaveBeenCalled()
    expect(props.fetchSettingsData).toHaveBeenCalled()
  })

  it('should not fetch data if public', () => {
    const root = shallow(<Bar {...props} isPublic={true} />)
    expect(props.fetchContext).not.toHaveBeenCalled()
    expect(props.fetchApps).not.toHaveBeenCalled()
    expect(props.fetchSettingsData).not.toHaveBeenCalled()
  })

  it('should not fetch apps if hasFetchedApps is true', () => {
    const root = shallow(<Bar {...props} hasFetchedApps={true} />)
    expect(props.fetchApps).toHaveBeenCalled()
  })

  it('should re-fetch apps if apps are not fetched and drawer opens', () => {
    const root = shallow(<Bar {...props} />)
    expect(props.fetchApps).toHaveBeenCalledTimes(1)
    root.setState({ drawerVisible: true })
    expect(props.fetchApps).toHaveBeenCalledTimes(2)
    root.setState({ drawerVisible: true, usageTracker: {} })
    expect(props.fetchApps).toHaveBeenCalledTimes(2)
  })

  it('should display the Searchbar', () => {
    const root = shallow(
      <Bar {...props} currentApp="Cozy Drive" isPublic={false} />
    )
    expect(toJson(root)).toMatchSnapshot()
  })

  it('should not display searchbar if we are on mobile', () => {
    isMobileApp.mockReturnValue(true)
    const root = shallow(
      <Bar {...props} currentApp="Cozy Drive" isPublic={false} />
    )
    expect(toJson(root)).toMatchSnapshot()
  })

  it('should not display searchbar if we are not on Cozy Drive', () => {
    isMobileApp.mockReturnValue(true)
    const root = shallow(
      <Bar {...props} currentApp="Cozy Contacts" isPublic={false} />
    )
    expect(toJson(root)).toMatchSnapshot()
  })

  it('should not display searchbar if we are not on  a public page', () => {
    isMobileApp.mockReturnValue(true)
    const root = shallow(
      <Bar {...props} currentApp="Cozy Drive" isPublic={true} />
    )
    expect(toJson(root)).toMatchSnapshot()
  })
})
