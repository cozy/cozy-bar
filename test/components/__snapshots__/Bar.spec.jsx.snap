// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bar should display the Searchbar 1`] = `
<div
  className="coz-bar-wrapper"
>
  <div
    className="coz-bar-container"
  >
    <button
      className="coz-bar-btn coz-bar-burger"
      data-icon="icon-apps"
      data-tutorial="apps-mobile"
      onClick={[Function]}
      type="button"
    >
      <span
        className="coz-bar-hidden"
      >
        drawer
      </span>
    </button>
    <Apps
      isPublic={false}
    />
    <div
      className="u-flex-grow"
    >
      <Wrapper />
    </div>
    <Wrapper
      toggleSupport={[Function]}
    />
    <Connect(Drawer)
      drawerListener={[Function]}
      isClaudyLoading={false}
      onClaudy={false}
      onClose={[Function]}
      toggleSupport={[Function]}
      visible={false}
    />
  </div>
</div>
`;

exports[`Bar should not display searchbar if we are not on  a public page 1`] = `
<div
  className="coz-bar-wrapper"
>
  <div
    className="coz-bar-container"
  >
    <Apps
      isPublic={true}
    />
    <div
      className="u-flex-grow"
    />
  </div>
</div>
`;

exports[`Bar should not display searchbar if we are not on Cozy Drive 1`] = `
<div
  className="coz-bar-wrapper"
>
  <div
    className="coz-bar-container"
  >
    <button
      className="coz-bar-btn coz-bar-burger"
      data-icon="icon-apps"
      data-tutorial="apps-mobile"
      onClick={[Function]}
      type="button"
    >
      <span
        className="coz-bar-hidden"
      >
        drawer
      </span>
    </button>
    <Apps
      isPublic={false}
    />
    <div
      className="u-flex-grow"
    />
    <Wrapper
      toggleSupport={[Function]}
    />
    <Connect(Drawer)
      drawerListener={[Function]}
      isClaudyLoading={false}
      onClaudy={false}
      onClose={[Function]}
      toggleSupport={[Function]}
      visible={false}
    />
  </div>
</div>
`;

exports[`Bar should not display searchbar if we are on mobile 1`] = `
<div
  className="coz-bar-wrapper"
>
  <div
    className="coz-bar-container"
  >
    <button
      className="coz-bar-btn coz-bar-burger"
      data-icon="icon-apps"
      data-tutorial="apps-mobile"
      onClick={[Function]}
      type="button"
    >
      <span
        className="coz-bar-hidden"
      >
        drawer
      </span>
    </button>
    <Apps
      isPublic={false}
    />
    <div
      className="u-flex-grow"
    />
    <Wrapper
      toggleSupport={[Function]}
    />
    <Connect(Drawer)
      drawerListener={[Function]}
      isClaudyLoading={false}
      onClaudy={false}
      onClose={[Function]}
      toggleSupport={[Function]}
      visible={false}
    />
  </div>
</div>
`;
